Feature/ruckus dpsk

julsemaan opened the pull request:
# Missing
* Docs (especially for the new unbound DPSK mode)

# Description
Adds support for bound and unbound DPSK on Ruckus.
Adds ability to reuse local account password for a DPSK
Adds a per-user VLAN pool (round robin that assigns a VLAN that is exclusive for a single user)

# Impacts
Ruckus module, DPSK flow, VLAN pooling

# NEW Package(s) required
perl(Crypt::PBKDF2), perl(Digest::SHA3)

# Delete branch after merge
YES

# Checklist
(REQUIRED) - [yes, no or n/a]
- [ ] Document the feature
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

# NEWS file entries
(REQUIRED, but may be optional...)
## New Features
* Added support for bound and unbound DPSK on Ruckus.
* Added ability to reuse local account password for a DPSK
* Added a per-user VLAN pool (round robin that assigns a VLAN that is exclusive for a single user)

jrouzierinverse commented on lib/pf/role/pool.pm:
```diff
@@ -180,6 +184,60 @@ sub getVlanByUsername {
+    WHERE  vlan IN ( $sql_vlans ) 
+           AND node.status = 'reg' 
+           AND pid = ? 
+    ", $pid);
```
Check if status to see if the query succeeded.
```perl
if (is_success($status)) {
```

jrouzierinverse replied:
Add a limit 1 to the SQL.

jrouzierinverse commented on lib/pf/role/pool.pm:
```diff
@@ -180,6 +184,60 @@ sub getVlanByUsername {
+           AND pid = ? 
+    ", $pid);
+
+    if(defined(my $row = $res->fetchrow_hashref)) {
```
Clean up the query.
```perl
$res->finish;
```

jrouzierinverse commented on lib/pf/role/pool.pm:
```diff
@@ -180,6 +184,60 @@ sub getVlanByUsername {
+        WHERE  vlan IN ( $sql_vlans ) 
+               AND node.status != 'unreg' 
+        ");
+        my %used_vlans = map{$_->[0] => 1} @{$res->fetchall_arrayref};
```
Clean up the query.
```perl
$res->finish;
```

jrouzierinverse commented on lib/pf/role/pool.pm:
```diff
@@ -180,6 +184,60 @@ sub getVlanByUsername {
+                    AND node.mac = locationlog.mac 
+        WHERE  vlan IN ( $sql_vlans ) 
+               AND node.status != 'unreg' 
+        ");
```
Check if status to see if the query succeeded.
```perl
if (is_success($status)) {
```

jrouzierinverse commented on html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm:
```diff
@@ -43,7 +43,7 @@ The main definition block
 has_block 'definition' =>
   (
-    render_list => [qw(id description root_module preregistration autoregister reuse_dot1x_credentials dot1x_recompute_role_from_portal mac_auth_recompute_role_from_portal dot1x_unset_on_unmatch dpsk default_psk_key unreg_on_acct_stop vlan_pool_technique)],
+    render_list => [qw(id description root_module preregistration autoregister reuse_dot1x_credentials dot1x_recompute_role_from_portal mac_auth_recompute_role_from_portal dot1x_unset_on_unmatch dpsk dpsk_sso default_psk_key unreg_on_acct_stop vlan_pool_technique)],
```
dpsk_sso? Is that a typo?

jrouzierinverse commented:
rebase

julsemaan commented:
just rebased and addressed the comments, I will not have access to a SZ to document this as it should. Still better to have the code in before it gets outdated again and I'll contact the customer to get some screenshots or see with @lzammit if we can revive the SZ at the office.

@jrouzierinverse, ready to merge if my changes look fine to you

jrouzierinverse commented on html/pfappserver/lib/pfappserver/Form/Config/Provisioning/dpsk.pm:
```diff
@@ -27,9 +27,14 @@ has_field 'psk_size' =>
 
+has_field 'dpsk_use_local_password' =>
+  (
+   type => 'Checkbox',
```
Use a toggle instead.
```perl
has_field 'dpsk_use_local_password' => (
   type => 'Toggle',
   checkbox_value => 'enabled',
   unchecked_value => 'disabled',
)
```

jrouzierinverse commented on html/pfappserver/lib/pfappserver/Form/Config/Provisioning/mobileconfig.pm:
```diff
@@ -74,6 +74,11 @@ has_field 'dpsk' =>
 
+has_field 'dpsk_use_local_password' =>
+  (
+   type => 'Checkbox',
```
Use Toggle instead.
```perl
has_field 'dpsk_use_local_password' => (
   type => 'Toggle',
   checkbox_value => 'enabled',
   unchecked_value => 'disabled',
)
```

jrouzierinverse commented on lib/pf/Switch.pm:
```diff
@@ -3590,8 +3590,9 @@ Set the current tenant in the DAL based on the tenant ID configured in the switc
-    pf::dal->set_tenant($self->{_TenantId});
+    my ($self, $radius_request) = @_;
+    my $tenant_id = $radius_request->{"PacketFence-Tenant-Id"} // $self->{_TenantId};
+    pf::dal->set_tenant($tenant_id);
```
Why would this be ever different?

julsemaan replied:
In my case I was using a RADIUS filter to map the same source switch IP to multiple tenants based on the content of the RADIUS request (like a SmartZone zone name). This PR adds all the plumbing so RADIUS filters can override the tenant ID when needed (both in auth+acct)

jrouzierinverse commented on lib/pf/Switch/Ruckus/SmartZone.pm:
```diff
@@ -233,6 +246,105 @@ sub deauthenticateMacWebservices {
+    my ($self, $radius_request) = @_;
+    my ($status, $iter) = pf::dal::person->search(
+        -where => {
+            psk => { '!=', '', '!=', undef },
```
That will not check for an empty string. Since it is overridden by the check for null.
Do this instead.
```perl
psk => {'!=' => [-and => '', undef]},
```

julsemaan commented:
adjustments done on the PR, please review again
